x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java

```diff
@@ -0,0 +1,563 @@
+
+    @Override
+    public boolean foldable() {
+        return left().foldable() && right().foldable();
+    }
+
+    @Override
+    public Object fold() {
+        return EvaluatorMapper.super.fold();
+    }
+
+    public abstract SpatialRelatesFunction withDocValues();
+
+    /**
+     * Push-down to Lucene is only possible if one field is an indexed spatial field, and the other is a constant spatial or string column.
+     */
+    public boolean canPushToSource(Predicate<FieldAttribute> isAggregatable) {
+        // The use of foldable here instead of SpatialEvaluatorFieldKey.isConstant is intentional to match the behavior of the
+        // Lucene pushdown code in EsqlTranslationHandler::SpatialRelatesTranslator
+        // We could enhance both places to support ReferenceAttributes that refer to constants, but that is a larger change
+        return isPushableFieldAttribute(left(), isAggregatable) && right().foldable()
+            || isPushableFieldAttribute(right(), isAggregatable) && left().foldable();
+    }
+
+    private static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAttribute> isAggregatable) {
+        return exp instanceof FieldAttribute fa
+            && fa.getExactInfo().hasExact()
+            && isAggregatable.test(fa)
+            && EsqlDataTypes.isSpatial(fa.dataType());
+    }
+
+    @Override
+    public int hashCode() {
+        // NB: the hashcode is currently used for key generation so
+        // to avoid clashes between aggs with the same arguments, add the class name as variation
+        return Objects.hash(getClass(), children(), useDocValues);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (super.equals(obj)) {
+            SpatialRelatesFunction other = (SpatialRelatesFunction) obj;
+            return Objects.equals(other.children(), children()) && Objects.equals(other.useDocValues, useDocValues);
+        }
+        return false;
+    }
+
+    public boolean useDocValues() {
+        return useDocValues;
+    }
+
+    /**
+     * Produce a map of rules defining combinations of incoming types to the evaluator factory that should be used.
+     */
+    protected abstract Map<SpatialEvaluatorKey, SpatialEvaluatorFactory<?, ?>> evaluatorRules();
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(
+        Function<Expression, EvalOperator.ExpressionEvaluator.Factory> toEvaluator
+    ) {
+        if (crsType == null) {
+            resolveType();
+        }
+        SpatialEvaluatorFactory<?, ?> factory = getSpatialEvaluatorFactory();
+        return factory.get(toEvaluator);
+    }
+
+    protected SpatialEvaluatorFactory<?, ?> getSpatialEvaluatorFactory() {
+        if (crsType == null) {
+            resolveType();
+        }
+        var evaluatorKey = new SpatialEvaluatorKey(crsType, useDocValues, fieldKey(left()), fieldKey(right()));
+        SpatialEvaluatorFactory<?, ?> factory = evaluatorRules().get(evaluatorKey);
```

Comment:
`evaluatorRules` should probably create a `static` map if we're going to use it like this.

Reply:
Done.